From 6f1d08f5513b0abb8a8300b389538414d444a36c Mon Sep 17 00:00:00 2001
From: "Subhransu S. Prusty" <subhransu.s.prusty@intel.com>
Date: Mon, 20 Oct 2014 13:58:01 +0530
Subject: [PATCH 1/1] ASoC: intel: Free the memcpy/sglist when clearing cached
 fw copy

During recovery there is memory leak because of not clearing the memcpy list
or sglist while we are clearing the cached copy of firmware.

Issue: GMINL-2585

Change-Id: I8003754ec8c03ba33e6912a0d92fae03e860f67e
Signed-off-by: Subhransu S. Prusty <subhransu.s.prusty@intel.com>
---
 sound/soc/intel/sst/sst_pvt.c |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/sound/soc/intel/sst/sst_pvt.c b/sound/soc/intel/sst/sst_pvt.c
index b327d91b6437..bdfb5f6896f1 100644
--- a/sound/soc/intel/sst/sst_pvt.c
+++ b/sound/soc/intel/sst/sst_pvt.c
@@ -429,6 +429,10 @@ void sst_do_recovery_mrfld(struct intel_sst_drv *sst)
 		pr_err("Clearing the cached FW copy...\n");
 		kfree(sst_drv_ctx->fw_in_mem);
 		sst_drv_ctx->fw_in_mem = NULL;
+		sst_memcpy_free_resources();
+		kfree(sst_drv_ctx->fw_sg_list.src);
+		kfree(sst_drv_ctx->fw_sg_list.dst);
+		sst_drv_ctx->fw_sg_list.list_len = 0;
 	}
 
 	mutex_lock(&sst->sst_lock);
-- 
1.7.9.5

